pulsar-package-manager/src/main/java/org/apache/pulsar/packages/manager/impl/PackageImpl.java

```diff
@@ -0,0 +1,146 @@
+        this.packageStorage = packageStorage;
+    }
+
+    @Override
+    public CompletableFuture<PackageMetadata> getMeta(PackageName packageName) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(exists -> exists
+                ? packageStorage.readAsync(metadataPath)
+                : FutureUtil.failedFuture(new PackageMetaNotFoundException("Package metadata does not existAsync")))
+            .thenApply(meta -> new Gson().fromJson(new String(meta), PackageMetadata.class));
+    }
+
+    @Override
+    public CompletableFuture<Void> updateMeta(PackageName packageName, PackageMetadata metadata) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(exists -> exists
+                ? packageStorage.writeAsync(metadataPath, new Gson().toJson(metadata).getBytes())
+                : FutureUtil.failedFuture(new PackageMetaNotFoundException("Package metadata does not existAsync")));
+    }
+
+    @Override
+    public CompletableFuture<Void> download(PackageName packageName, OutputStream outputStream) {
+        String pacakgePath = getPackageStoragePath(packageName);
+        return packageStorage.existAsync(pacakgePath)
+            .thenCompose(exists -> exists
+                ? packageStorage.readAsync(pacakgePath, outputStream)
+                : FutureUtil.failedFuture(new PackageNotFoundException("Package does not existAsync")));
+    }
+
+    @Override
+    public CompletableFuture<Void> upload(PackageName packageName, PackageMetadata metadata, InputStream inputStream) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        String packagePath = getPackageStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(metaExists -> metaExists
+                ? FutureUtil.failedFuture(new PackageMetaAlreadyExistsException("Package metadata already exists"))
+                : packageStorage.existAsync(packagePath))
+            .thenCompose(dataExists -> dataExists
+                ? FutureUtil.failedFuture(new PackageAlreadyExistsException("Package already exists"))
+                : packageStorage.writeAsync(metadataPath, new Gson().toJson(metadata).getBytes()))
+            .thenCompose(ignore -> packageStorage.writeAsync(packagePath, inputStream));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(PackageName packageName) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        String packagePath = getPackageStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(metaExists -> metaExists
+                ? packageStorage.deleteAsync(metadataPath)
+                : FutureUtil.failedFuture(new PackageMetaNotFoundException("Package metadata does not exists")))
```

Comment:
data can still exists when the metadata file doesn't exist. e.g. a delete operation failed in between deleting metadat and data.